### PR TITLE
role namespaces, workflows

### PR DESCRIPTION
**Pull request by kylesferrazza**
- [x] associate exam with a BN course
- [x] split routes and controllers based on roles
- [x] show exams with relevant links on homepage
- [x] remove unused gems and npm packages
- [x] restructure app/javascript to match new routes

### CODE REVIEW

**Comment by kylesferrazza** [comment]
- [ ] professor/exams#create
- [ ] professor/exams#new
- [ ] professor/exams#edit
- [ ] professor/exams#show
- [ ] professor/exams#update
- [ ] professor/exams#update
- [ ] proctor/exams#finalize
- [ ] proctor/anomalies#index
- [ ] proctor/anomalies#show
- [ ] proctor/anomalies#destroy
- [ ] proctor/registrations#clear_anomalies
- [ ] proctor/registrations#finalize
- [ ] proctor/rooms#finalize
- [ ] proctor/rooms#index
- [ ] proctor/rooms#show
- [ ] proctor/exams#show
- [ ] student/exams#start
- [ ] student/exams#submit
- [ ] student/anomalies#create
- [ ] student/exams#save_snapshot
- [ ] student/exams#ask_question
- [ ] student/exams#show

**Comment by kylesferrazza** [comment]
Still need to:
- [ ] move displaying an exam to `proctor/`
- [ ] make `student/` follow routes

File: app/controllers/main_controller.rb [hunk 1]
```diff
@@ -1,9 +1,36 @@
...
       redirect_to new_user_session_path
+      return
     end
+
+    registrations = current_user.registrations
+    exams =
+      registrations
+      .map do |r|
+        e = r.exam
+        return nil unless e.enabled?
```
Comment (blerner) [inline app/controllers/main_controller.rb:13]: This `return` will bail all the way out of `index`, not just this loop iteration.

File: app/controllers/professor/exams_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    @courses = semesters.map do |s|
+      s['courses'].collect do |c|
+        [c['name'], c['id']]
+      end
+    end.flatten(1)
+    pp @courses
+  end
+
+  def create
+    # TODO make sure prof is registered for the course in BN before creating
```
Comment (blerner) [inline app/controllers/professor/exams_controller.rb:21]: Let's be sure to do this before merging this PR.

File: app/views/proctor/registrations/show.html.erb [hunk 3]
```diff
@@ -7,14 +7,14 @@
       <% if @registration.anomalous? %>
         <p><%= link_to "Show anomalies", exam_registration_anomalies_path(@registration.exam, @registration), class: "btn btn-danger" %></p>
       <% end %>
-      <p><%= link_to "View exam registrations", exam_registrations_path(@registration.exam), class: "btn btn-info" %></p>
-      <p><%= link_to "View exam", @registration.exam, class: "btn btn-info" %></p>
-      <p><%= link_to "View room", [@exam, @registration.room], class: "btn btn-info" %></p>
+      <p><%= link_to "View exam registrations", proctor_exam_registrations_path(@registration.exam), class: "btn btn-info" %></p>
+      <p><%= link_to "View exam", ['proctor', @registration.exam], class: "btn btn-info" %></p>
+      <p><%= link_to "View room", ['proctor', @exam, @registration.room], class: "btn btn-info" %></p>
```
Comment (blerner) [inline app/views/proctor/registrations/show.html.erb:12]: It's weird to me that these three `link_to`'s aren't using the same syntax; chalk it up to Rails, I guess?

File: app/controllers/professor/exams_controller.rb [hunk 4]
```diff
@@ -14,19 +14,25 @@ def new
...
       end
     end.flatten(1)
-    pp @courses
   end
 
   def create
-    # TODO make sure prof is registered for the course in BN before creating
     exam_params = params.require(:exam).permit(:name, :file, :course_id, :enabled)
+    course_id = exam_params[:course_id]
+    courses = bottlenose_token.get('/api/courses').parsed
```
Comment (blerner) [inline app/controllers/professor/exams_controller.rb:22]: This could fail, and the uncaught exception would lead to a Err500

Comment (blerner) [reply app/controllers/professor/exams_controller.rb:22]: (In general, we need to be really careful about using the `bottlenose_token` in an unguarded manner)

Comment (kylesferrazza) [reply app/controllers/professor/exams_controller.rb:22]: Yes, we should probably catch this at the application level and display an error about contacting bottlenose?

Comment (kylesferrazza) [reply app/controllers/professor/exams_controller.rb:22]: #88

**Pull request merged**